### PR TITLE
Fix pod IP downward API test flake

### LINKED ISSUE #20403: e2e flake: Downward API should provide pod IP as an env var
http://kubekins.dls.corp.google.com/view/Critical%20Builds/job/kubernetes-e2e-gce/10735/

```
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/downward_api.go:81 "POD_IP=(?:\\d+)\\.(?:\\d+)\\.(?:\\d+)\\.(?:\\d+)" in container output Expected : POD_IP= KUBERNETES_PORT=tcp://10.0.0.1:443 KUBERNETES_SERVICE_PORT=443 HOSTNAME=downward-api-b0eb2988-c8ab-11e5-afda-42010af01555 SHLVL=1 HOME=/root KUBERNETES_PORT_443_TCP_ADDR=10.0.0.1 PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin KUBERNETES_PORT_443_TCP_PORT=443 KUBERNETES_PORT_443_TCP_PROTO=tcp KUBERNETES_PORT_443_TCP=tcp://10.0.0.1:443 KUBERNETES_SERVICE_PORT_HTTPS=443 PWD=/ KUBERNETES_SERVICE_HOST=10.0.0.1 to match regular expression : POD_IP=(?:\d+)\.(?:\d+)\.(?:\d+)\.(?:\d+)
```

I'm not sure who's responsible for Downward API. cc @lavalamp @davidopp @bgrant0607

### PR DESCRIPTION
**Pull request by pmorie**
Fixes #20403; still needs unit coverage.

### CODE REVIEW

**Comment by thockin** [comment]
Looks deceptively simple.  Can you please explain here or in the associated bug what happened and what the fix is?

Ping me when tests are in, and we can rush this

**Comment by pmorie** [comment]
@thockin it was deceptively simple -- basically we made a very foolish assumption about where to source the pod IP -- the old code had it jury-rigged up so that the first SyncPod call for the pod in the dockertools Manager would populate the pod IP on that `*api.Pod`s status, and then we would read the pod IP from that field in `makeEnvironmentVariables`.  That is a terrible assumption to make, though, and I believe there's a race where a container might be started and assigned an IP but an inspect call to the daemon may not return it.  So, between the race and our naive decisions about assuming the status would always be populated in the pod, you got flakes.

File: pkg/kubelet/kubelet.go [hunk 1]
```diff
@@ -1486,7 +1486,14 @@ func (kl *Kubelet) podFieldSelectorRuntimeValue(fs *api.ObjectFieldSelector, pod
...
 	switch internalFieldPath {
 	case "status.podIP":
-		return pod.Status.PodIP, nil
+		status, _ := kl.statusManager.GetPodStatus(pod.UID)
+		if status.PodIP != "" {
+			return status.PodIP, nil
+		}
+
+		glog.V(4).Infof("Calling out to container runtime for status of pod %v/%v in order to inject pod IP", pod.Namespace, pod.Name)
+		runtimeStatus, _ := kl.containerRuntime.GetPodStatus(pod.UID, pod.Name, pod.Namespace)
```
Comment (vishh) [inline pkg/kubelet/kubelet.go]: Is the container guaranteed to be running at this point?

Comment (pmorie) [inline pkg/kubelet/kubelet.go]: @vishh, the infra container should be running for docker pods.  We will need to rethink this for rkt, but that's obviously not at issue here.

Comment (vishh) [inline pkg/kubelet/kubelet.go]: *should* or *will*? You are ignoring the error code here and that's the reason for my comment.

Comment (pmorie) [inline pkg/kubelet/kubelet.go]: that was not intentional, probably sleep deprivation at work; i wll go back over this and fix

**Comment by thockin** [comment]
Is this PR done?  Stalled? Dead?

**Comment by pmorie** [comment]
@thockin @saad-ali updated

**Comment by saad-ali** [comment]
LGTM

**Comment by ncdc** [comment]
@k8s-bot test this github flake: #21755

**Comment by ncdc** [comment]
@pmorie I think there's a chance that this PR might have caused #21755. I can repro that issue with your branch.

File: pkg/kubelet/dockertools/manager.go [hunk 2]
```diff
@@ -2052,6 +2050,9 @@ func (dm *DockerManager) GetPodStatus(uid types.UID, name, namespace string) (*k
 		}
 		containerStatuses = append(containerStatuses, result)
 		if ip != "" {
+			if podStatus.IP != ip {
```
Comment (vishh) [inline pkg/kubelet/dockertools/manager.go]: Add some comments on ordering here @pmorie! We don't want a refactoring to screw up this logic later.

**Comment by ncdc** [comment]
Removed lgtm because 1 unit test is broken and the changes here broke /etc/hosts

File: pkg/kubelet/dockertools/manager.go (hunk 2 above)
Comment (ncdc) [inline pkg/kubelet/dockertools/manager.go]: @vishh this is just a logging addition, right?

Comment (vishh) [inline pkg/kubelet/dockertools/manager.go]: My bad. I actually meant that comment for this PR in general. I will move my comment to `kubelet.go` changes.

File: pkg/kubelet/kubelet.go [hunk 3]
```diff
@@ -1500,7 +1500,18 @@ func (kl *Kubelet) podFieldSelectorRuntimeValue(fs *api.ObjectFieldSelector, pod
 	}
 	switch internalFieldPath {
 	case "status.podIP":
-		return pod.Status.PodIP, nil
+		status, _ := kl.statusManager.GetPodStatus(pod.UID)
```
Comment (vishh) [inline pkg/kubelet/kubelet.go]: Add some comments on ordering here @pmorie! We don't want a refactoring to screw up this logic later.

**Comment by pmorie** [comment]
Closing, superceded by #22666

**Pull request closed without merging**